src/drivers/amd/rembrandt/acp_dmic_dma.c

```diff
@@ -0,0 +1,377 @@
+
+static uint32_t dmic_rngbuff_size;
+/* 109c7aba-a7ba-43c3-b9-42-59-e2-0a-66-11-be */
+DECLARE_SOF_UUID("acp_dmic_dma", acp_dmic_dma_uuid, 0x109c7aba, 0xa7ba, 0x43c3,
+		0xb9, 0x42, 0x59, 0xe2, 0x0a, 0x66, 0x11, 0xbe);
+DECLARE_TR_CTX(acp_dmic_dma_tr, SOF_UUID(acp_dmic_dma_uuid), LOG_LEVEL_INFO);
+
+/* allocate next free DMA channel */
+static struct dma_chan_data *acp_dmic_dma_channel_get(struct dma *dma,
+						   unsigned int req_chan)
+{
+	k_spinlock_key_t key;
+	struct dma_chan_data *channel;
+
+	key = k_spin_lock(&dma->lock);
+	if (req_chan >= dma->plat_data.channels) {
+		k_spin_unlock(&dma->lock, key);
+		tr_err(&acp_dmic_dma_tr, "ACP_DMIC: Channel %d out of range",
+				req_chan);
+		return NULL;
+	}
+	channel = &dma->chan[req_chan];
+	if (channel->status != COMP_STATE_INIT) {
+		k_spin_unlock(&dma->lock, key);
+		tr_err(&acp_dmic_dma_tr, "ACP_DMIC: Cannot reuse channel %d",
+				req_chan);
+		return NULL;
+	}
+	atomic_add(&dma->num_channels_busy, 1);
+	channel->status = COMP_STATE_READY;
+	k_spin_unlock(&dma->lock, key);
+	return channel;
+}
+
+static void acp_dmic_dma_channel_put(struct dma_chan_data *channel)
+{
+	k_spinlock_key_t key;
+
+	notifier_unregister_all(NULL, channel);
+	key = k_spin_lock(&channel->dma->lock);
+	channel->status = COMP_STATE_INIT;
+	atomic_sub(&channel->dma->num_channels_busy, 1);
+	k_spin_unlock(&channel->dma->lock, key);
+}
+
+static int acp_dmic_dma_start(struct dma_chan_data *channel)
+{
+	acp_wov_pdm_no_of_channels_t pdm_channels;
+	acp_wov_pdm_decimation_factor_t deci_fctr;
+	acp_wov_misc_ctrl_t wov_misc_ctrl;
+	acp_wov_pdm_dma_enable_t  pdm_dma_enable;
+	struct timer *timer = timer_get();
+	uint64_t deadline = platform_timer_get(timer) +
+				clock_ms_to_ticks(PLATFORM_DEFAULT_CLOCK, 1) * 500 / 1000;
+	channel->status = COMP_STATE_ACTIVE;
+	if (channel->direction == DMA_DIR_DEV_TO_MEM) {
+		/* Channel for DMIC */
+		pdm_channels.bits.pdm_no_of_channels = 0;
+		io_reg_write(PU_REGISTER_BASE + ACP_WOV_PDM_NO_OF_CHANNELS,
+							pdm_channels.u32all);
+		/* Decimation Factor */
+		deci_fctr.u32all = 2;
+		io_reg_write(PU_REGISTER_BASE + ACP_WOV_PDM_DECIMATION_FACTOR,
+							deci_fctr.u32all);
+		/* PDM Control */
+		wov_misc_ctrl = (acp_wov_misc_ctrl_t)
+			io_reg_read(PU_REGISTER_BASE + ACP_WOV_MISC_CTRL);
+		wov_misc_ctrl.u32all |= 0x10;
+		io_reg_write(PU_REGISTER_BASE + ACP_WOV_MISC_CTRL,
+						wov_misc_ctrl.u32all);
+		/* PDM Enable */
+		io_reg_write(PU_REGISTER_BASE + ACP_WOV_PDM_ENABLE, 1);
+		/* PDM DMA Enable */
+		io_reg_write(PU_REGISTER_BASE + ACP_WOV_PDM_DMA_ENABLE, 1);
+		/* Check the PDM DMA Enable status bit */
+		pdm_dma_enable = (acp_wov_pdm_dma_enable_t)
+			io_reg_read(PU_REGISTER_BASE + ACP_WOV_PDM_DMA_ENABLE);
+		while (!(uint32_t)pdm_dma_enable.bits.pdm_dma_en_status) {
+			if (deadline < platform_timer_get(timer)) {
+				/* safe check in case we've got preempted after read */
+				if ((uint32_t)pdm_dma_enable.bits.pdm_dma_en_status)
+					return 0;
```

Comment:
This is a safe check to avoid indefinite while loop if hw went to bad state and not returning expected status.
before entering to loop we read status register and wait for status to be succeeded. if hw is not return success then we wait for some  specified time and return error.

 /* Check the PDM DMA Enable status bit */
 pdm_dma_enable = (acp_wov_pdm_dma_enable_t)
                        io_reg_read(PU_REGISTER_BASE + ACP_WOV_PDM_DMA_ENABLE);
                while (!(uint32_t)pdm_dma_enable.bits.pdm_dma_en_status) {
                        if (deadline < platform_timer_get(timer)) {
                                /* safe check in case we've got preempted after read */
                                if ((uint32_t)pdm_dma_enable.bits.pdm_dma_en_status)
                                        return 0;
                                tr_err(&acp_dmic_dma_tr, "DMICDMA: timed out for dma start");
                                return -ETIME;
                        }
                        pdm_dma_enable = (acp_wov_pdm_dma_enable_t)
                                        io_reg_read(PU_REGISTER_BASE + ACP_WOV_PDM_DMA_ENABLE);
                }
above is the complete code for loop.

Reply:
> ```c
>  /* Check the PDM DMA Enable status bit */
>  pdm_dma_enable = (acp_wov_pdm_dma_enable_t)
>                         io_reg_read(PU_REGISTER_BASE + ACP_WOV_PDM_DMA_ENABLE);
>                 while (!(uint32_t)pdm_dma_enable.bits.pdm_dma_en_status) {
> ```

Answering my above comment - I see we are reading the HW here so the result is valid for the loop above.

Reply:
pdm_dma_enable  variable is mapped to read PDM DMA register. So this variable will be updated based hw status

Reply:
I don't think it's "mapped" to hw registers. It's a perfectly standard function-local object. You *read* hardware registers into it before the loop and at the bottom of the loop. Between `while` and this second `if` the object cannot change.